### DISCUSSION TITLE
Question on Conditional Controls

### DISCUSSION THREAD

**Discussion by pengfejeac** [idaholab/moose#27100] [Q&A Modules: General] [answered]
Greetings!

I'm working on a problem that requires switching between conditions for more than one time and I'm kinda confused about which control/function should I apply. The rough idea is as follows:

Let's say I apply "HeatConduction" kernels/bcs to describe the temperature of an object and what I wish to do is: When the object's overall temperature reaches 50 degree, I will turn off the heater object(or boundaries) and when the temperature drops below 0 degree, I will turn the heater back on. The switching process will repeat itself for a couple of times.

I'm aware that there's a control object called "ConditionalEnableControl", but it seems like it will switch condition for only once, which is not suitable for my case. I'm perfectly fine with creating my own kernels/objects in Moose if that's necessary.

Any help will be much appreciated!

**Comment by GiudGiud** [comment]
Hello

You should use a conditional control object for that.
Bake the logic with the two thresholds in a ParsedFunction 
https://mooseframework.inl.gov/moose/source/controls/ConditionalFunctionEnableControl.html

Guillaume

**Reply by pengfejeac** [reply]
I see, thanks! So if I'm switching values of heat source, I can use Real function control. But if I'm switching between different conditions, for example, Dirichlet BC and Neumann BC, then I will have to use Bool function control right? 

By the way, could Parsed Function take values from post processor?

**Reply by GiudGiud** [reply]
Hello

Yes, to change the value you ll want to use this control
https://mooseframework.inl.gov/moose/source/controls/RealFunctionControl.html
to change the status of 'active/inactive' of an object, you ll want to use this one
https://mooseframework.inl.gov/moose/source/controls/ConditionalFunctionEnableControl.html

ParsedFunction can accept postprocessors, with the symbol_names / values parameters
https://mooseframework.inl.gov/moose/source/functions/MooseParsedFunction.html

**Reply by pengfejeac** [reply]
That works! Thanks a lot for your help! I just have another quick question on performance of such Control actions: I tend to use Iterative solver in most of the cases and when the problem is well conditioned and well constructed, "dt" could get very large. So the "turning point"(the moment to enable/disable stuffs) of Control action will not be very accurate, do you have any idea on how to fix this issue other than using fixed dt throughout the process?

**Reply by GiudGiud** [reply, marked as answer]
Hello

So you can execute controls within a time step, on nonlinear iterations. I am not sure I would recommend doing so, as this could cause the nonlinear solve to 'ping pong' between active and inactive on some objects.

I think the clean way to do this would be to use a Terminator to fail a time step (with some additional logic to allow it if dt < 1s for example) when the condition for the Control switch occurs.
https://mooseframework.inl.gov/source/userobjects/Terminator.html

Guillaume

**Reply by pengfejeac** [reply]
Terminator with a soft stop sounds like a very reasonable way to do this, I'll give it a shot! Thanks again for your help! I really appreciate it.
